### PR TITLE
Rebin

### PR DESCRIPTION
**Pull request by jwaterfield**
Function to rebin spectra. Adapted from http://wiki.scipy.org/Cookbook/Rebinning

### CODE REVIEW

**Comment by jwaterfield** [comment]
Noticed a couple of bugs sorry. Will be fixed in next commit

File: echidna/core/spectra.py [hunk 1]
```diff
@@ -229,3 +229,29 @@ def add(self, spectrum):
...
+
+        Raises:
+          ValueError: Number of dimensions incorrect for spectra
+          ValueError: Old bins/ New bins must be integer
+        '''
+        shape = self._data.shape
+        len_shape = len(shape)
+        if len_shape != len(args):
+            raise ValueError("Number of dimensions incorrect")
+        for i in range(len_shape):
```
Comment (ashleyrback) [inline echidna/core/spectra.py]: Again variable name could be a little bit more descriptive e.g. `dim`.

File: echidna/core/spectra.py [hunk 2]
```diff
@@ -229,3 +229,29 @@ def add(self, spectrum):
...
+            if shape[i] % args[i] != 0:
+                raise ValueError("Old bins/ New bins must be integer")
+        factor = numpy.asarray(shape)/numpy.asarray(args)
+        evList = ['self._data.reshape('] + \
+            ['args[ % d],factor[%d],' % (i, i) for i in range(len_shape)] + \
+            [')'] + ['.sum(%d)' % (i+1) for i in range(len_shape)] + \
+            ['/factor[ % d]' % i for i in range(len_shape)]
+        self._data = eval(''.join(evList))
```
Comment (ashleyrback) [inline echidna/core/spectra.py]: I think after `self._data` has been rebinned, the relevant members that describe the binning should be updated. I.e. if `self._data` is rebinned along the energy axis, then `self._energy_bins` and `self._energy_width` should be updated accordingly.

File: echidna/scripts/rebin.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    parser.add_argument("-b", "--bins", nargs='+', type=int,
+                        help="Number of bins for each dimension")
+    args = parser.parse_args()
+    directory = args.input[:args.input.rfind("/")+1]  # strip filename
+    # strip directory and extension
+    filename = args.input[args.input.rfind("/")+1:args.input.rfind(".")]
+    spectrum = store.load(args.input)
+    print spectrum._data.shape
+    cmd = 'spectrum.rebin('
+    for b in args.bins:
```
Comment (ashleyrback) [inline echidna/scripts/rebin.py]: Variable name could be a little more descriptive. Looping over dimensions so e.g. `dim`

File: echidna/core/spectra.py [hunk 4]
```diff
@@ -229,3 +229,29 @@ def add(self, spectrum):
         self._data += spectrum._data
         self._raw_events += spectrum._raw_events
         self._num_decays += spectrum._num_decays
+
+    def rebin(self, *args):
+        '''rebin spectra data into a smaller spectra of the same rank whose
+        dimensions are factors of the original dimensions.
+        Adapted from http://wiki.scipy.org/Cookbook/Rebinning
```
Comment (ashleyrback) [inline echidna/core/spectra.py]: Could you split this into two paragraphs, a short sentence for the first line and then a longer sentence with more explanation. Also capitalise "Rebin"

File: echidna/scripts/rebin.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
+""" Example smearing script
+
+This script:
+  * Reads in spectrum from hdf5
+  * Rebins spectrum with user input number of bins
+  * Saves rebinned spectrum to given output or, by default, to the same
+    directory with _rebin added to the filename.
```
Comment (ashleyrback) [inline echidna/scripts/rebin.py]: Sorry I'm being a bit picky but could you either end all bullet points with a full-stop or none (unless the bullet has multiple sentences) not mix and match.

**Pull request merged**